server/forge/bitbucketdatacenter/bitbucketdatacenter.go

```diff
@@ -0,0 +1,637 @@
+	SkipVerify bool
+	Consumer   *oauth.Consumer
+}
+
+// New returns a Forge implementation that integrates with Bitbucket DataCenter/Server,
+// the on-premise edition of Bitbucket Cloud, formerly known as Stash.
+func New(opts Opts) (forge.Forge, error) {
+	config := &client{
+		url:        opts.URL,
+		URLApi:     fmt.Sprintf("%s/rest", opts.URL),
+		Username:   opts.Username,
+		Password:   opts.Password,
+		SkipVerify: opts.SkipVerify,
+	}
+
+	switch {
+	case opts.Username == "":
+		return nil, fmt.Errorf("Must have a git machine account username")
+	case opts.Password == "":
+		return nil, fmt.Errorf("Must have a git machine account password")
+	case opts.ConsumerKey == "":
+		return nil, fmt.Errorf("Must have a oauth1 consumer key")
+	}
+
+	if opts.ConsumerRSA == "" && opts.ConsumerRSAString == "" {
+		return nil, fmt.Errorf("must have CONSUMER_RSA_KEY set to the path of a oauth1 consumer key file or CONSUMER_RSA_KEY_STRING set to the value of a oauth1 consumer key")
+	}
+
+	var keyFileBytes []byte
+	if opts.ConsumerRSA != "" {
+		var err error
+		keyFileBytes, err = os.ReadFile(opts.ConsumerRSA)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		keyFileBytes = []byte(opts.ConsumerRSAString)
+	}
+
+	block, _ := pem.Decode(keyFileBytes)
+	PrivateKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	config.Consumer = CreateConsumer(opts.URL, opts.ConsumerKey, PrivateKey)
+	return config, nil
+}
+
+// Name returns the string name of this driver
+func (c *client) Name() string {
+	return "bitbucket_dc"
+}
+
+// URL returns the root url of a configured forge
+func (c *client) URL() string {
+	return c.url
+}
+
+func (c *client) Login(ctx context.Context, res http.ResponseWriter, req *http.Request) (*model.User, error) {
+	requestToken, u, err := c.Consumer.GetRequestTokenAndUrl("oob")
+	if err != nil {
+		return nil, err
+	}
+	code := req.FormValue("oauth_verifier")
+	if len(code) == 0 {
+		http.Redirect(res, req, u, http.StatusSeeOther)
+		return nil, nil
+	}
+	requestToken.Token = req.FormValue("oauth_token")
+	accessToken, err := c.Consumer.AuthorizeToken(requestToken, code)
+	if err != nil {
+		return nil, err
+	}
+
+	client := internal.NewClientWithToken(ctx, c.url, c.Consumer, accessToken.Token)
+	userID, err := client.FindCurrentUser()
+	if err != nil {
+		return nil, err
+	}
+	userID = strings.ReplaceAll(userID, "@", "_") // Apparently the "whoami" endpoint may return the wrong username
+
+	bc, err := c.newClient(&model.User{Token: accessToken.Token})
+	if err != nil {
+		return nil, fmt.Errorf("unable to create bitbucket client: %w", err)
+	}
+
+	user, _, err := bc.Users.GetUser(ctx, userID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to query for user: %w", err)
+	}
+
+	return convertUser(user, accessToken.Token, c.url), nil
+}
+
+// Auth is not supported.
+func (*client) Auth(_ context.Context, _, _ string) (string, error) {
+	return "", fmt.Errorf("Not Implemented")
```

Comment:
Could you explain the oauth flow in the case for BB-DC. Normally `Login()` is used to redirect the user to the forge and `Auth` is used for the callback.